FixWhitespace.hs

```diff
@@ -125,50 +126,48 @@ main = do
       verbose = optVerbose opts
       config  = optConfig  opts
 
+  base <- getCurrentDirectory
+
   files <- if not $ null nonOpts
-    then concat <$> traverse namesMatching nonOpts
+    then getDirectoryFiles base nonOpts
     else do
-      Config incDirs excDirs incFiles excFiles <- parseConfig config
-      base <- getCurrentDirectory
-      find (validDir base incDirs excDirs) (validFile base incFiles excFiles) base
+      Config incDirs0 excDirs0 incFiles excFiles <- parseConfig config
+      let incDirs = map (++ "/**/") incDirs0
+      let excDirs = map (++ "/**/") excDirs0
+
+      -- File patterns to always include
+      -- when not matching an excluded file pattern
+      let incWhitelistPatterns = concatMap (\d -> map (d ++) incFiles) incDirs
+      -- File patterns to always exclude
+      let excBlacklistPatterns = map ("**/" ++) excFiles
+
+      -- Files to include when not in an excluded directory
+      -- and when not matching an excluded file pattern
+      let incPatterns = map ("**/" ++) incFiles
+      -- Directory and file patterns to exclude
+      let excPatterns = (map (++ "*") excDirs)
+                     ++ (map ("**/" ++) excFiles)
+
+      when verbose (do putStrLn "Include whitelist:"
+                       putStrLn (concatMap (++ "\n") incWhitelistPatterns)
+
+                       putStrLn "Exclude blacklist:"
+                       putStrLn (concatMap (++ "\n") excBlacklistPatterns)
+
+                       putStrLn "Include:"
+                       putStrLn (concatMap (++ "\n") incPatterns)
+
+                       putStrLn "Exclude:"
+                       putStrLn (concatMap (++ "\n") excPatterns))
+
+      files0 <- getDirectoryFilesIgnore base incWhitelistPatterns excBlacklistPatterns
+      files1 <- getDirectoryFilesIgnore base incPatterns excPatterns
+      return (nub (files0 ++ files1))
```

Comment:
Use `nubOrd` from the `extra` package instead?
http://neilmitchell.blogspot.com/2015/02/nub-considered-harmful.html
```haskell
import Data.List.Extra (nubOrd)
```

Reply:
Didn't know about that - looks good!